SSA/Experimental/IntrinsicAsymptotics.lean

```diff
@@ -93,47 +99,344 @@ mk_ex 120
+match pattern with
+    | .var v => (lets, getVarAfterMapping v m)
+    | .add a b => 
+      let (lets, v1) := applyMapping a m lets
+      let (lets, v2) := applyMapping b m lets
+      let l := lets.size - v1 - 1
+      let r := lets.size - v2 - 1
+      (lets.push (Expr.add l r), lets.size)
+    | .cst n => (lets.push (.cst n), lets.size)
+
+def shiftBy (inputProg : Com) (delta: Nat) (pos: Nat := 0): Com := 
+  let shift (v : Nat) :=
+    if v >= pos then
+      v + delta
+    else
+      v
+  match inputProg with
+  | .ret x => .ret (shift x)
+  | .let ty e body => match e with
+    | .add a b => .let ty (Expr.add (shift a) (shift b)) (shiftBy body delta (pos+1))
+    | .cst x => .let ty (.cst x) (shiftBy body delta (pos +1))
+
+def replaceUsesOfVar (inputProg : Com) (old: Nat) (new : Var) : Com := 
+  let replace (v : Nat) :=
+     if old = v then new else v
+  match inputProg with
+  | .ret x => .ret (replace x)
+  | .let ty e body => match e with
+    | .add a b => 
+      .let ty (Expr.add (replace a) (replace b)) (replaceUsesOfVar body (old+1) (new+1))
+    | .cst x => .let ty (.cst x) (replaceUsesOfVar body (old+1) (new+1))
+
+def addLetsToProgram' (newLets : Lets) (n : Nat) (newProgram : Com) : Com :=
+  match n with
+  | 0 => newProgram
+  | (x + 1) => if x < newLets.size then
+                 let added := Com.let .nat (newLets[x]!) newProgram
+                 addLetsToProgram' newLets x (added)
+               else
+                 newProgram
+
+def addLetsToProgram (newLets : Lets) (newProgram : Com) : Com :=
+  addLetsToProgram' newLets newLets.size newProgram
+
+def applyRewrite (lets : Lets) (inputProg : Com) (rewrite: ExprRec × ExprRec) : Option Com := do
+  let varPos := lets.size - 1 
+  let mapping ← matchVar lets varPos rewrite.1
+  let (newLets, newVar) := applyMapping (rewrite.2) mapping lets
+  let newProgram := inputProg
+  let newProgram := shiftBy newProgram (newLets.size - lets.size)
+  let newProgram := replaceUsesOfVar newProgram (newLets.size - lets.size) (newLets.size - newVar - 1)
+  let newProgram := addLetsToProgram newLets newProgram
+
+  some newProgram
+
+def rewriteAt' (inputProg : Com) (depth : Nat) (lets: Lets) (rewrite: ExprRec × ExprRec) : Option Com :=
+  match inputProg with
+    | .ret _ => none
+    | .let _ expr body =>
+        let lets := lets.push expr
+        if depth = 0 then
+           applyRewrite lets body rewrite
+        else
+           rewriteAt' body (depth - 1) lets rewrite
+
+def rewriteAt (inputProg : Com) (depth : Nat) (rewrite: ExprRec × ExprRec) : Option Com :=
+    rewriteAt' inputProg depth #[] rewrite 
+
+def rewrite (inputProg : Com) (depth : Nat) (rewrite: ExprRec × ExprRec) : Com :=
+    let x := rewriteAt inputProg depth rewrite 
+    match x with
+      | none => inputProg
+      | some y => y
 
 def Expr.denote : Expr → State → Value
-| .nat n, _ => .nat n
+| .cst n, _ => .nat n
 | .add a b, l => let a_val := get_nat (l.get! a)
```

Comment:
I wonder if it might be better to use `getD` here, instead of `get!`. For the theorems, I think `get!` with an out of bounds index will default to `0` anyway, so we might as well lean into it and state that all uninitialized variables get implicitly initialized to `0`.

I would also be a fan of indexing `Expr` with an upper-bound on the number of free variables, and making `State` be backed by a `Vector`. Mostly because I like modelling these things exactly, I don't know for sure if it would actually make our lives easier.